swaylock/password.c

```diff
@@ -0,0 +1,125 @@
+#include <assert.h>
+#include <pwd.h>
+#include <security/pam_appl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <wlr/util/log.h>
+#include <xkbcommon/xkbcommon.h>
+#include "swaylock/swaylock.h"
+#include "swaylock/seat.h"
+#include "unicode.h"
+
+static int function_conversation(int num_msg, const struct pam_message **msg,
+		struct pam_response **resp, void *data) {
+	struct swaylock_password *pw = data;
+	/* PAM expects an array of responses, one for each message */
+	struct pam_response *pam_reply = calloc(
+			num_msg, sizeof(struct pam_response));
+	*resp = pam_reply;
+	for (int i = 0; i < num_msg; ++i) {
+		switch (msg[i]->msg_style) {
+		case PAM_PROMPT_ECHO_OFF:
+		case PAM_PROMPT_ECHO_ON:
+			pam_reply[i].resp = pw->buffer;
+			break;
+		case PAM_ERROR_MSG:
+		case PAM_TEXT_INFO:
+			break;
+		}
+	}
+	return PAM_SUCCESS;
+}
+
+static bool attempt_password(struct swaylock_password *pw) {
+	struct passwd *passwd = getpwuid(getuid());
+	char *username = passwd->pw_name;
+	const struct pam_conv local_conversation = {
+		function_conversation, pw
+	};
+	pam_handle_t *local_auth_handle = NULL;
+	int pam_err;
+	if ((pam_err = pam_start("swaylock", username,
+					&local_conversation, &local_auth_handle)) != PAM_SUCCESS) {
+		wlr_log(L_ERROR, "PAM returned error %d", pam_err);
+	}
+	if ((pam_err = pam_authenticate(local_auth_handle, 0)) != PAM_SUCCESS) {
+		wlr_log(L_ERROR, "pam_authenticate failed");
+		goto fail;
+	}
+	if ((pam_err = pam_end(local_auth_handle, pam_err)) != PAM_SUCCESS) {
+		wlr_log(L_ERROR, "pam_end failed");
+		goto fail;
+	}
+	// PAM frees this
+	pw->buffer = NULL;
+	pw->len = pw->size = 0;
+	return true;
+fail:
+	// PAM frees this
+	pw->buffer = NULL;
+	pw->len = pw->size = 0;
+	return false;
+}
+
+static bool backspace(struct swaylock_password *pw) {
+	if (pw->len != 0) {
+		pw->buffer[--pw->len] = 0;
```

Comment:
Side note: with assignments in `if`s, side effects in array brackets are practices I don't like and are error-prone. Especially for a lockscreen.

Reply:
I don't think this is error prone, this behavior is well defined

Reply:
I'm not saying this is undefined behavior, I'm just saying it's unreadable and thus error-prone, even for people used to these idioms. It's just like `if`s without brackets.

Anyway, I don't care too much, this is just a warning to say "if there's a bug because of this, don't blame me for having approved this PR".

Reply:
I know what you're saying. I disagree.